### PR TITLE
change: Add more validations for pipeline step new interfaces

### PR DESCRIPTION
**Pull request by qidewenwhen**
*Description of changes:* Add more validations for pipeline step new interfaces to prevent misusage.

*Testing done:* Unit tests

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/workflow/model_step.py [hunk 1]
```diff
@@ -57,17 +57,52 @@ def __init__(
...
+                    dict(
+                       register_model_retry_policies=[
+                           StepRetryPolicy(...)
+                       ],
+                       repack_model_retry_policies=[
+                           SageMakerJobStepRetryPolicy(...)
+                       ],
+                    )
+
+                OR
```
[the remark this answers is not in the document]
Comment (qidewenwhen) [reply src/sagemaker/workflow/model_step.py:78]: Updated in the new commit

File: src/sagemaker/workflow/model_step.py [hunk 2]
```diff
@@ -57,17 +57,52 @@ def __init__(
                 If a listed `Step` name does not exist, an error is returned (default: None).
             retry_policies (List[RetryPolicy] or Dict[str, List[RetryPolicy]]): The list of retry
                 policies for the `ModelStep` (default: None).
+                If a list of retry policies is provided, it would be applied to all steps in the
+                ModelStep collection. Note: in this case, `SageMakerJobStepRetryPolicy`
+                is not allowed as create/register model step does not support it.
+                If a dict is provided, it can specify different retry policies for different
+                types of steps in the ModelStep collection. Similarly, `SageMakerJobStepRetryPolicy`
+                is not allowed for create/register model step. See examples below:
```
[the remark this answers is not in the document]
Comment (qidewenwhen) [reply src/sagemaker/workflow/model_step.py:65]: Updated in the new commit

File: src/sagemaker/workflow/pipeline_context.py [hunk 3]
```diff
@@ -15,13 +15,44 @@
...
+        """Create a `_StepArguments`
+
+        Args:
+            caller_name (str): The name of the caller function which is intercepted by the
+                PipelineSession to get the step arguments.
+        """
+        self.caller_name = caller_name
+
+
+class _RegularStepArguments(_StepArguments):
```
[the remark this answers is not in the document]
Comment (qidewenwhen) [reply src/sagemaker/workflow/pipeline_context.py:36]: hh, yea, good point! Updated in the new commit

File: src/sagemaker/estimator.py [hunk 4]
```diff
@@ -1318,6 +1318,11 @@ def register(
         Returns:
             str: A string of SageMaker Model Package ARN.
         """
+        if isinstance(self.sagemaker_session, PipelineSession):
+            raise TypeError(
+                "estimator.register does not support PipelineSession at this moment. "
+                "Please use model.register under PipelineSession if you're using the ModelStep."
```
Comment (jerrypeng7773) [inline src/sagemaker/estimator.py:1324]: nit: Please use `model.register` with `PipelineSession` ?

Comment (qidewenwhen) [reply src/sagemaker/estimator.py:1324]: Updated in the new commit

**Review by navaj0 — APPROVED** [review]

**Review by jerrypeng7773 — APPROVED** [review]

**Review by navinsoni — APPROVED** [review]

**Review by nmadan — APPROVED** [review]

**Review by shreyapandit — APPROVED** [review]

**Pull request merged**